src/Features/Lsif/Generator/Program.cs

```diff
@@ -264,11 +266,10 @@ private static async Task LocateAndRegisterMSBuild(TextWriter logFile, Directory
             await logFile.WriteLineAsync($"Load of the binlog complete; {msbuildInvocations.Length} invocations were found.");
 
             var lsifGenerator = Generator.CreateAndWriteCapabilitiesVertex(lsifWriter, logFile);
+            var workspace = new AdhocWorkspace(await Composition.CreateHostServicesAsync());
 
             foreach (var msbuildInvocation in msbuildInvocations)
```

Comment:
I'm curious if there's any benefit in parallelizing this step at all?

Reply:
My general assumption was proejcts tend to have enough documents to probably saturate. We could do this in parallel but it'd also mean additional memory usage because now we're holding multiple projects at once versus just the one we're analyzing.

Reply:
in FindRefs we found that processing a single project at a time (ideally in topological order) and then preocessing in parallel *within* the project produced the best results.

Reply:
FWIW I downloaded this new version and ingested the VS Editor binlog and it looks like it's averaging just one or two CPU cores of activity at a time.

I didn't drill much deeper than peeking at Task Manager but I thought this was interesting as I'd assume this is mostly CPU bound work?

Reply:
It occasionally bursts higher, but I wonder if this points to a bottleneck, like that lock around STDOUT, that could be improved.

Reply:
Best would be an actual trace.

Reply:
I would doubt it's locking (unless the trace shows otherwise). If it is locking, we should move away from streaming the results. That serves no purpose. This should be a map/reduce (or fork join) that operates on files in parallel, and writes the result in one op for a project when all docs are complete.

Reply:
FWIW the current model of a single gigantic LSIF is very costly on the consumer side as well because we can only read a graph on a single thread at a time and end up holding all monikers it covers in memory until the end: https://github.com/microsoft/language-server-protocol/issues/1755

In conversation with Dirk, it sounds like the preferred strategy might be to emit an LSIF per project (or other generator decided unit).

I'm curious if that dramatically improves throughput on the production side as well.

Reply:
Does the consumer side matter though?  How does that impact the end to end experience?  It would be nice to get a breakdown off costs here. I would not have expected consumption to be costly since everything is computed up front. 

And yes, lsif perf project is the only granularity that makes sense to me.

Reply:
For example, the same file will be in multiple projects, with different semantics. So you'd need to know which project you were looking through to make the meaning of a particular file be correct.

Reply:
> Does the consumer side matter though?

It impacts server costs, read duration, scalability (how big of a workspace we can handle before we run out of memory on that machine), reliability (probability of an OOM exception).

> I would not have expected consumption to be costly since everything is computed up front.

LSIF is a graph so one has to hold much of it in memory at a time to be able to reason about it. It can also only really be traversed by a single thread at a time, so large LSIFs mean machines consuming LSIF have a single core pinned at 100%, the rest mostly idle, and much of their RAM utilized, and can't really process any other LSIFs simultaneously unless you can be sure they are all small enough to fit in the remaining RAM.

There are ways to work-around this by allocating and [tracking RAM dedicated to each in-progress LSIF ingestion](https://devdiv.visualstudio.com/DevDiv/_git/VSCloudKernel/pullrequest/484772) but this is more complicated and it'd be preferable to have more, smaller, LSIFs.

With small, predictable sizes, the level of concurrency is more consistent and OOMs consuming large or unusual workspaces will de facto never happen.

> you'd need to know which project you were looking through to make the meaning of a particular file be correct

I think this already works today. AFAIK the LSIF generator creates a separate document node for each project that contains the document. This data is encoded via the `contains` edge, which indicates that the document is contained by that specified project.

Reply:
Why do you have to process with a single thread?  It's all immutable data. I would expect parallel processing to be trivial. After all, that's what we can do in the production side. It seems odd here that consumption would have it harder. 

Furthermore, I would expect deep parallelization in servers due to processing a queue of lsif files. When I worked at G, that's how things worked. It wasn't like you spun up a server to process one piece of data. You had servers dequeuing what was essentially an unbounded queue of work. So they were always saturated. 

Finally, if lsif isn't a good format for processing, then we should create one that is.  My presumption on using lsif as the intermediary format was that it was actually good for this. If not, we can absolutely come up with formats that are tuned for low resource usage, and trivial algorithms on both construction and ingestion. :-)

Reply:
> Finally, if lsif isn't a good format for processing, then we should create one that is. My presumption on using lsif as the intermediary format was that it was actually good for this. If not, we can absolutely come up with formats that are tuned for low resource usage, and trivial algorithms on both construction and ingestion. :-)

That's the gist of [the issue](https://github.com/microsoft/language-server-protocol/issues/1755) I linked. It describes incremental changes to make it more efficient to process.

Reply:
Put another way. This should be like game development. We should think about our formats first, optimizing for performance. We should not start everything with bad structures (like graphs or json) and then have to contort to make that decent :-)

In the end, the product we are shipping is not "lsif support", it's "rich code experiences in the web". We should start with that and move backwards. If that indicates lsif is a poor choice because of design and cogs, we can make something else.